### ISSUE TITLE
[Potential contribution] 'AdaFamily' optimizer

### ISSUE THREAD

**Issue by hfassold** [microsoft/DeepSpeed#1810]
I would like to contribute the new 'AdaFamily' optimizer to DeepSpeed.

AdaFamily is a novel method for training deep neural networks. It is a family of adaptive gradient methods and can be interpreted as sort of a blend of the state-of-the-art optimization algorithms Adam, AdaBelief and AdaMomentum. We perform experiments on standard datasets for image classification, demonstrating that our proposed method outperforms these algorithms.

The AdaFamily algorithm  and the experiments with computer vision (image classification) tasks are described in the preprint at  https://arxiv.org/abs/2203.01603 (paper has been submitted to ISPR 2022).

Code for the AdaFamily algorithm and other SoA optimizers (AdaBelief, AdaMomentum, Adam etc.) can be found at https://github.com/hfassold/omni_optimizer .

After submitting the paper, according to the DeepSpeed contribution guidelines which demand also experiments with NLP tasks, I did additionally an experiment with a NLP finetuning task. Code for the experiments is at https://github.com/hfassold/nlp_finetuning_adafamily . We do a NLP finetuning of a pretrained 'DistillBert' model on the task of 'text classification' (prediction of google app rating von 1-5). We train for 10 epochs, and eps is set to 1e-16. We compare with the baseline optimizer 'AdamW'. As performance measures, we take the average accuracy (over 10 runs with different seeds). AdaFamily is used with myu set to 0.25 . AdamW achieves an average accuracy of 79.94 percent, whereas AdaFamily achieves an average accuracy of 80.47 percent. So AdaFamily is ~ 0.5 percent better here.

**Comment by conglongli** [comment]
Hi @hfassold , we read your proposal and paper, and we feel that there are a few concerns that make it difficult for us to accept the contribution:
(1) The most concern we have is that based on the current experiment results it's hard to conclude whether AdaFamily really provides nontrivial convergence gain. In your paper (https://arxiv.org/pdf/2203.01603.pdf) section 3, for AdaFamily you tuned its hyperparameter µ, while other algorithms have no tuning. And only one combination of LR/beta1/beta2/weight decay/etc. is applied to all algorithms. So it's natural to question whether additional tuning of existing algorithms will lead to the same or better accuracy than AdaFamily, especially given that AdaFamily doesn't provide a huge accuracy increase in Table 1-3. I personally think at least we need to try different LR for all algorithms since AdaFamily's additional hyperparameter µ affects the effective LR.

(2) AdaFamily has no runtime/system performance benefit compared to baseline algorithms. This makes it less attractive to DeepSpeed users considering the concern in (1).

(3) There is no theoretical convergence analysis.

Finally, I noticed that at the end of your paper it says "In the future, we will port the AdaFamily algorithm to a distributed training framework like DeepSpeed.". We would appreciate if you could refrain from including description like this before a confirmed code release in DeepSpeed, to avoid potential confusion to paper readers and DeepSpeed users. Thank you. And in fact in the paper I didn't find any reference to your own code release (https://github.com/hfassold/omni_optimizer), which I feel is much more important to include.

**Comment by hfassold** [comment]
Hi, thanks for your investigation ! You are right, the optimization of the hyperparameters was limited. Actually, I optimized the hyperparameters (LR schedule etc.) for the 'Adam' algorithm, and kept it then the same for all other algorithms.

**Comment by conglongli** [comment]
> Actually, I optimized the hyperparameters (LR schedule etc.) for the 'Adam' algorithm, and kept it then the same for all other algorithms.

I think the issue here is that due to each algorithm's different variance calculation, even if you set the LR hyperparameter the same, the effective LR (where variance is part of it) is different. Thus different algorithms may have different "optimal LR hyperparameter". So a better comparison would be: try a wide set of different LR for all algorithms, and each algorithm chooses its own LR that produces the best result, and finally compare the best results between algorithms.

**Comment by conglongli** [comment]
Closing this as we decided to not take this contribution as discussed above, but users can always plugin any customized optimizer when using DeepSpeed.
